### ISSUE TITLE
3D particle emitter improvements

### ISSUE THREAD

**Issue by jack-ii** [GDevelopApp/GDevelop-extensions#1110]
### Is there an existing issue for this?

- [X] I have searched the [existing issues](https://github.com/GDevelopApp/GDevelop-extensions/issues)

### Enter the name of the extension

3D particle emitter

### Describe the bug of the extension

- [ ] The Hide/Show actions dont work. fix or remove
Also the available object-effect and size actions have no result.

- [ ] Swap the position of End & Start color/opacity in the properties so that it is more intuitive and consistent with the 2D emitter

- [ ] Move the Z position from object properties window into properties side panel, like 3D objects
Then 3D-Emitters are not overlapped by every 3D object in the scene-editor and it is also possible for instances to have a different z position without doing it with events

- [ ] Add (0 = infinitely) to the Emission duration property and action

- [ ] Particle emitters placed in the scene are always active for a moment after the scene starts
(0.0001 Duration or 0 Flow doesn't help)
Maybe an Enable checkbox and action would be possible. Or 0 is 0 and a checkbox/action for endless duration

- [ ] I am unable to create a dark gray smoke because these colors become invisible, even without a light layer, 255 opacity, and dark particle image

- [ ] Add 3D particle emitter to GD
The extension code is big and will get bigger in the future with more premade effects and is already available in the New object list without the extension being installed.

Because both extensions are from the same dev, two things about CurvedMovement(speed-based)
- [ ] Allow to move backwards without flipping the sprite/3D-object
If I use a minus value in the acceleration action to drive a car/train backwards, it flips automatically so that the front side is always forward. Maybe a checkbox to not do this or an action to change the Rotation offset in the behavior.

- [ ] Swap the position of Acceleration and Targeted speed fields in der Acceleration action, to be more intuitive and fit the eventsheet order 
![01](https://github.com/GDevelopApp/GDevelop-extensions/assets/81538128/fdb08bba-fe05-4f47-b6bb-c7d8cd37fe9f)
![02](https://github.com/GDevelopApp/GDevelop-extensions/assets/81538128/41d0830d-1dc9-4225-b8ad-e15560c6a444)

### Steps to reproduce

.

### GDevelop platform

Desktop

### GDevelop version

5.3.181

**Comment by D8H** [comment]
Thanks for the feedback.

> 3D particle emitter
> ### Describe the bug of the extension
> 
> * [ ]  The Hide/Show actions dont work. fix or remove
>   Also the available object-effect and size actions have no result.

Particles rendering is batched by the library I use so the default hide implementation doesn't work. I could try to override it but it may cause issues in the future and I doubt hiding particle is really useful.

> * [ ]  Swap the position of End & Start color/opacity in the properties so that it is more intuitive and consistent with the 2D emitter

This is a limitation of the community extension system. It sorts alphabetically.

> * [ ]  Move the Z position from object properties window into properties side panel, like 3D objects
>   Then 3D-Emitters are not overlapped by every 3D object in the scene-editor and it is also possible for instances to have a different z position without doing it with events

Custom objects were made for 2D. It's a limitation of the community extension system.

> * [ ]  Add (0 = infinitely) to the Emission duration property and action

I will at some point. The workaround is to put a big value. For instance, 123456 seconds = 35 hours.

> * [ ]  Particle emitters placed in the scene are always active for a moment after the scene starts
>   (0.0001 Duration or 0 Flow doesn't help)
>   Maybe an Enable checkbox and action would be possible. Or 0 is 0 and a checkbox/action for endless duration

Interesting, I should try to fix this.

> * [ ]  I am unable to create a dark gray smoke because these colors become invisible, even without a light layer, 255 opacity, and dark particle image

I'm not sure to understand. Is it because of the blending property is set on additive? I guess that additive and black is just adding 0.

> * [ ]  Add 3D particle emitter to GD
>   The extension code is big and will get bigger in the future with more premade effects and is already available in the New object list without the extension being installed.

The extension code size is not really how built-in extension are chosen but the code comes from a library so it already allows very advanced features and only 5% is actually used.

> Because both extensions are from the same dev, two things about CurvedMovement(speed-based)
> 
> * [ ]  Allow to move backwards without flipping the sprite/3D-object
>   If I use a minus value in the acceleration action to drive a car/train backwards, it flips automatically so that the front side is always forward. Maybe a checkbox to not do this or an action to change the Rotation offset in the behavior.

I though it was already the case. In the examples, the train can go backward like a train.

> * [ ]  Swap the position of Acceleration and Targeted speed fields in der Acceleration action, to be more intuitive and fit the eventsheet order

I'm not sure if it would be more intuitive or not, but it probably not worth deprecating the action just to change the parameter order.

**Comment by jack-ii** [comment]
Thanks for the quick and good response.

> I will at some point. The workaround is to put a big value.

I meant add the text "0 = infinitely" in the action description. 0 already causes endless running time.

> Is it because of the blending property is set on additive?

I think that's it, thank you.

> The extension code size is not really how built-in extension are chosen but the code comes from a library so it already allows very advanced features and only 5% is actually used.

It seems a bit random which features are built in and which are extensions. My feeling was that if they are more complex and are not needed very rarely, they should be built in. To avoid the limitation of the extension system as in this case.

> I though it was already the case. In the examples, the train can go backward like a train.

In Train(isometry) yes, in Train(top-down) not. I don't really see what the difference is, but I'll try again.


> probably not worth deprecating the action just to change the parameter order.

I'm not sure what that means, but it looks like it takes two minutes and has no disadvantages.

**Comment by D8H** [comment]
> I'm not sure what that means, but it looks like it takes two minutes and has no disadvantages.

It would be quick to change it in the extension but it would be a new action and the old one would be hidden. When you edit an action and it doesn't select anything in the action list, it's because it was hidden (deprecated).

**Comment by D8H** [comment]
> In Train(isometry) yes, in Train(top-down) not. I don't really see what the difference is, but I'll try again.

Indeed, the isometric train doesn't rotate, it change its animation. I guess the rotation can be done manually according to the movement angle and the speed but a property would probably be better.
